sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCacheImpl.java

```diff
@@ -0,0 +1,169 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+import reactor.core.publisher.Sinks;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+public class AccessTokenCacheImpl {
+    // The delay after a refresh to attempt another token refresh
+    private static final Duration REFRESH_DELAY = Duration.ofSeconds(30);
+    // the offset before token expiry to attempt proactive token refresh
+    private static final Duration REFRESH_OFFSET = Duration.ofMinutes(5);
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
+    private final AtomicReference<Sinks.One<AccessToken>> wip;
+    private final TokenCredential tokenCredential;
+    private TokenRequestContext tokenRequestContext;
+    private final Predicate<AccessToken> shouldRefresh;
+    private final ClientLogger logger = new ClientLogger(AccessTokenCacheImpl.class);
+
+    /**
+     * Creates an instance of AccessTokenCacheImpl with default scheme "Bearer".
+     *
+     * @param tokenCredential the credential to be used to acquire token from.
+     */
+    public AccessTokenCacheImpl(TokenCredential tokenCredential) {
+        Objects.requireNonNull(tokenCredential, "The token credential cannot be null");
+        this.wip = new AtomicReference<>();
+        this.tokenCredential = tokenCredential;
+        this.shouldRefresh = accessToken -> OffsetDateTime.now()
+            .isAfter(accessToken.getExpiresAt().minus(REFRESH_OFFSET));
+    }
+
+    /**
+     * Asynchronously get a token from either the cache or replenish the cache with a new token.
+     *
+     * @param tokenRequestContext The request context for token acquisition.
+     * @return The Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> getToken(TokenRequestContext tokenRequestContext) {
+        return Mono.defer(retrieveToken(tokenRequestContext))
+            // Keep resubscribing as long as Mono.defer [token acquisition] emits empty().
+            .repeatWhenEmpty((Flux<Long> longFlux) -> longFlux.concatMap(ignored -> Flux.just(true)));
+    }
+
+    private Supplier<Mono<? extends AccessToken>> retrieveToken(TokenRequestContext tokenRequestContext) {
+        return () -> {
+            try {
+                if (wip.compareAndSet(null, Sinks.one())) {
+                    final Sinks.One<AccessToken> sinksOne = wip.get();
+                    OffsetDateTime now = OffsetDateTime.now();
+                    Mono<AccessToken> tokenRefresh;
+                    Mono<AccessToken> fallback;
+
+                    Supplier<Mono<AccessToken>> tokenSupplier = () ->
```

Comment:
Could you further document why there is a class level `tokenRequestContext` and a parameter level instance as well.